Libraries/Components/MaskedView/MaskedViewIOS.ios.js

```diff
@@ -0,0 +1,170 @@
+   */
+  renderMask: () => React.Element<*>,
+};
+
+type State = {
+  maskViewNodeRef: ?mixed,
+};
+
+/**
+ * Renders the child view with a mask specified in the `renderMask` prop.
+ *
+ * ```
+ * import React from 'react';
+ * import { MaskedView, Text, View } from 'react-native';
+ *
+ * class MyMaskedView extends React.Component {
+ *   render() {
+ *     return (
+ *       <MaskedView
+ *         style={{ flex: 1 }}
+ *         renderMask={() =>
+ *           <View style={styles.maskContainerStyle}>
+ *             <Text style={styles.maskTextStyle}>
+ *               Basic Mask
+ *             </Text>
+ *           </View>
+ *         }
+ *       >
+ *         <View style={{ flex: 1, backgroundColor: 'blue' }} />
+ *       </MaskedView>
+ *     );
+ *   }
+ * }
+ * ```
+ *
+ * The above example will render a view with a blue background that fills its
+ * parent, and then mask that view with text that says "Basic Mask".
+ *
+ * The alpha channel of the view rendered by the `renderMask` prop determines how
+ * much of the view’s content and background shows through. Fully or partially
+ * opaque pixels allow the underlying content to show through but fully
+ * transparent pixels block that content.
+ *
+ */
+class MaskedViewIOS extends React.Component {
+  props: Props;
+
+  state: State = {
+    maskViewNodeRef: null,
+  };
+
+  static propTypes = {
+    ...ViewPropTypes,
+    renderMask: ReactPropTypes.func.isRequired,
+  };
+
+  _hasWarnedInvalidChildren = false;
+  _hasWarnedInvalidRenderMask = false;
+  _maskViewNodeHandle: ?number = null;
+
+  componentWillUpdate(nextProps: Props, nextState: State) {
+    if (nextState.maskViewNodeRef !== this.state.maskViewNodeRef) {
+      this._maskViewNodeHandle = findNodeHandle(nextState.maskViewNodeRef);
+    }
+  }
+
+  render() {
+    if (!this.props.children || React.Children.count(this.props.children) > 1) {
+      if (!this._hasWarnedInvalidChildren) {
+        console.warn(
+          'MaskedView: MaskedView requires a single child React Element.'
+        );
+        this._hasWarnedInvalidChildren = true;
+      }
+      return null;
+    }
+
+    const target = React.Children.only(this.props.children);
+
+    if (typeof this.props.renderMask !== 'function') {
+      if (!this._hasWarnedInvalidRenderMask) {
+        console.warn(
+          'MaskedView: Invalid `renderMask` prop was passed to MaskedView. ' +
+            'Expected a function. No mask will render.'
+        );
+        this._hasWarnedInvalidRenderMask = true;
+      }
+      return <View style={this.props.style}>{target}</View>;
+    }
+
+    const maskElement = this.props.renderMask();
+    let maskElementWithRef = null;
+    if (maskElement) {
+      maskElementWithRef = cloneReferencedElement(maskElement, {
+        ref: this._handleMaskViewRef,
+      });
+    }
+
+    return (
+      <RCTMaskedView
+        style={this.props.style}
+        maskRef={this._maskViewNodeHandle}>
+        <View
```

Comment:
The reason I wrapped the mask in a view is so that I didn't have to do anything weird to get Yoga to lay it out appropriately. If I just create the wrapper view on the native side, layout doesn't occur as I want it to.

However, I see the value of simplifying the JS side -- so if you have thoughts on how to make sure the mask is still laid out properly while creating it's container natively, I'm happy to do that. It just wasn't clear to me how to make that happen, and this seemed like the best way.